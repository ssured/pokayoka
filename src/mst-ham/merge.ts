import { entries } from 'mobx';

export type THam = { [key: string]: number | [number, THam] };
export type HamValue = THam['any'];
const EMPTY_HAM_STATE = -1;

const clone = <T extends any>(o: T): T => {
  const s = JSON.stringify(o);
  return s ? JSON.parse(s) : undefined;
};

export const isObject = (x: any): x is object =>
  (typeof x === 'object' && x !== null) || typeof x === 'function';

function stateFromHam(ham: HamValue | undefined): number {
  return ham === undefined ? 0 : typeof ham === 'number' ? ham : ham[0];
}

export function merge(
  machineState: number,
  incomingHam: HamValue,
  incomingValue: any,
  currentHam: HamValue,
  currentValue: any
): {
  resultHam: HamValue;
  resultValue: any;
  currentChanged: boolean;
  deferUntilState?: number;
} {
  const incomingState = stateFromHam(incomingHam);

  // console.log(machineState);
  // console.log(incomingState);
  // console.log(incomingValue);
  // console.log(stateFromHam(currentHam));
  // console.log(currentValue);

  if (machineState < incomingState) {
    // defer
    // the incoming value is outside the boundary of the machine's state, it must be reprocessed in another state.
    return {
      resultHam: currentHam,
      resultValue: currentValue,
      currentChanged: false,
      deferUntilState: incomingState,
    };
  }

  const currentState = stateFromHam(currentHam);

  if (incomingState < currentState) {
    // historical
    // the incoming value is within the boundary of the machine's state, but not within the range.
    return {
      resultHam: currentHam,
      resultValue: currentValue,
      currentChanged: false,
    };
  }

  // if incoming is an object, compare deeper
  // objects are more important than primitives
  if (Array.isArray(incomingHam) && isObject(incomingValue)) {
    const currentIsObject = isObject(currentValue);

    const resultHam: [number, THam] = currentIsObject
      ? typeof currentHam === 'number'
        ? [
            currentHam,
            Object.keys(currentValue).reduce(
              (map, key) => {
                map[key] = currentHam;
                return map;
              },
              {} as THam
            ),
          ]
        : clone(currentHam as [number, THam])
      : [0, {}];

    const replace = stateFromHam(incomingHam) > stateFromHam(currentHam);
    if (replace) resultHam[0] = stateFromHam(incomingHam);

    const currentObject = currentIsObject ? currentValue : {};
    const resultValue = clone(currentObject);

    let currentChanged = false;
    let deferUntilState: number | undefined = undefined;

    for (const [key, incomingSubHam] of entries(incomingHam[1])) {
      const incomingSubValue = (incomingValue as any)[key];

      const currentSubHam = replace ? EMPTY_HAM_STATE : resultHam[1][key];
      const currentSubValue = replace ? undefined : currentObject[key];

      // console.log(incomingSubHam);
      // console.log(incomingSubValue);
      // console.log(currentSubHam);
      // console.log(currentSubValue);

      const subMerge = merge(
        machineState,
        incomingSubHam,
        incomingSubValue,
        currentSubHam,
        currentSubValue
      );

      // console.log(subMerge.resultHam);
      // console.log(subMerge.resultValue);

      if (subMerge.resultHam !== EMPTY_HAM_STATE) {
        resultHam[1][key] = subMerge.resultHam;
      }

      if (subMerge.resultValue === undefined) {
        if (key in resultValue) {
          delete resultValue[key];
        }
      } else {
        resultValue[key] = subMerge.resultValue;
      }

      currentChanged = currentChanged || subMerge.currentChanged;
      if (subMerge.deferUntilState != null) {
        deferUntilState = Math.min(
          deferUntilState || Infinity,
          subMerge.deferUntilState
        );
      }
    }

    return deferUntilState == null
      ? { resultHam, resultValue, currentChanged }
      : { resultHam, resultValue, currentChanged, deferUntilState };
  }

  if (currentState < incomingState) {
    // incoming value is newer
    return {
      resultHam: incomingHam,
      resultValue: incomingValue,
      currentChanged: true,
    };
  }

  // currentState === incomingState
  if (incomingValue === currentValue) {
    // equal
    return {
      resultHam: currentHam,
      resultValue: currentValue,
      currentChanged: false,
    };
  }

  // finally just JSON compare and select the winner
  // prefer objects over primitive, otherwise compare JSON lexed value
  const currIsObj = isObject(currentValue);
  const incIsObj = isObject(incomingValue);

  return (currIsObj && !incIsObj) ||
    (!currIsObj &&
      !incIsObj &&
      (JSON.stringify(incomingValue) || '') <
        (JSON.stringify(currentValue) || ''))
    ? {
        resultHam: currentHam,
        resultValue: currentValue,
        currentChanged: false,
      }
    : {
        resultHam: incomingHam,
        resultValue: incomingValue,
        currentChanged: true,
      };
}
