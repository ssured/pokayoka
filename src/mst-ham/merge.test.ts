import { merge, HamValue } from './merge';

describe('ham objects merge', () => {
  test('it merges simple objects', () => {
    expect(
      merge(3, [2, { a: 2 }], { a: 'A' }, [1, { a: 1 }], { a: 'a' })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: 'A' },
      currentChanged: true,
    });

    // order does not matter
    expect(
      merge(3, [1, { a: 1 }], { a: 'a' }, [2, { a: 2 }], { a: 'A' })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: 'A' },
      currentChanged: false,
    });

    // defer if in future
    expect(
      merge(1, [2, { a: 2 }], { a: 'A' }, [1, { a: 1 }], { a: 'a' })
    ).toEqual({
      resultHam: [1, { a: 1 }],
      resultValue: { a: 'a' },
      currentChanged: false,
      deferUntilState: 2,
    });

    // adding properties
    expect(
      merge(4, [1, { b: 3 }], { b: 'b' }, [1, { a: 2 }], { a: 'a' })
    ).toEqual({
      resultHam: [1, { a: 2, b: 3 }],
      resultValue: { a: 'a', b: 'b' },
      currentChanged: true,
    });

    // equal
    expect(
      merge(4, [1, { a: 2 }], { a: 'a' }, [1, { a: 2 }], { a: 'a' })
    ).toEqual({
      resultHam: [1, { a: 2 }],
      resultValue: { a: 'a' },
      currentChanged: false,
    });
  });

  test('it prefers larger value when states are equal and type is equal', () => {
    expect(
      merge(3, [2, { a: 2 }], { a: 'z' }, [2, { a: 2 }], { a: 'a' })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: 'z' },
      currentChanged: true,
    });

    expect(
      merge(3, [2, { a: 2 }], { a: 'a' }, [2, { a: 2 }], { a: 'z' })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: 'z' },
      currentChanged: false,
    });

    expect(
      merge(3, [2, { a: 2 }], { a: 'a' }, [2, { a: 2 }], { a: undefined })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: 'a' },
      currentChanged: true,
    });
    expect(
      merge(3, [2, { a: 2 }], { a: undefined }, [2, { a: 2 }], { a: 'a' })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: 'a' },
      currentChanged: false,
    });
  });

  test('it prefers objects over primitives', () => {
    expect(
      merge(3, [2, { a: 2 }], { a: { b: 'B' } }, [2, { a: 2 }], { a: 'a' })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: { b: 'B' } },
      currentChanged: true,
    });

    expect(
      merge(3, [2, { a: 2 }], { a: 'a' }, [2, { a: 2 }], { a: { b: 'B' } })
    ).toEqual({
      resultHam: [2, { a: 2 }],
      resultValue: { a: { b: 'B' } },
      currentChanged: false,
    });
  });

  test('it merges nested objects', () => {
    expect(
      merge(
        3,
        [1, { sub: [2, { a: 2 }] }],
        { sub: { a: 'A' } },
        [1, { a: 1 }],
        { a: 'a' }
      )
    ).toEqual({
      resultHam: [1, { a: 1, sub: [2, { a: 2 }] }],
      resultValue: { a: 'a', sub: { a: 'A' } },
      currentChanged: true,
    });
  });

  test('it merges nested objects, and notifies of future values', () => {
    expect(
      merge(
        3,
        [1, { sub: [2, { a: 2, b: 4 }] }],
        { sub: { a: 'A', b: 'future value' } },
        [1, { a: 1 }],
        { a: 'a' }
      )
    ).toEqual({
      resultHam: [1, { a: 1, sub: [2, { a: 2 }] }],
      resultValue: { a: 'a', sub: { a: 'A' } },
      currentChanged: true,
      deferUntilState: 4,
    });
  });

  test('returns a defer state if only partial merge happened', () => {
    expect(
      merge(2, [1, { b: 2, c: 3 }], { b: 2, c: 3 }, [1, { a: 1 }], { a: 1 })
    ).toEqual({
      resultHam: [1, { a: 1, b: 2 }],
      resultValue: { a: 1, b: 2 },
      currentChanged: true,
      deferUntilState: 3,
    });
  });

  test('it stores deleted keys of nested objects', () => {
    expect(
      merge(
        4,
        [1, { sub: [2, { a: 2, b: 3 }] }],
        { sub: { a: 'A' } },
        [1, { sub: [2, { a: 2, b: 2 }] }],
        { sub: { a: 'A', b: 'B' } }
      )
    ).toEqual({
      resultHam: [1, { sub: [2, { a: 2, b: 3 }] }],
      resultValue: { sub: { a: 'A' } },
      currentChanged: true,
    });
  });
});

describe('real world usage', () => {
  test('something with the _rev not appearing in the result value', () => {
    const inSnap = {
      _id: 'cisvmclrpxin',
      _rev: '3045-132fec4e1aac785e74ac884ae0da5dac',
      '#': [
        1543234836586,
        {
          type: 1543234836586,
          _id: 1543234836586,
          name: 1543243812705,
          plans: [1543246278801, { cit4g6b5xqop: 1543246278801 }],
        },
      ] as HamValue,
      type: 'project',
      name: '1234',
      plans: {
        cit4g6b5xqop: {
          type: 'plan',
          _id: 'cit4g6b5xqop',
          name: 'plan 1543246278800',
        },
      },
    };
    const curSnap = {
      type: 'project',
      _id: 'cisvmclrpxin',
      '#': [
        1543234836586,
        {
          type: 1543234836586,
          _id: 1543234836586,
          name: 1543243812705,
          plans: [1543246278801, { cit4g6b5xqop: 1543246278801 }],
        },
      ] as HamValue,
      name: '1234',
      plans: {
        cit4g6b5xqop: {
          type: 'plan',
          _id: 'cit4g6b5xqop',
          name: 'plan 1543246278800',
        },
      },
    };

    const result = merge(
      Date.now(),
      inSnap['#'],
      inSnap,
      curSnap['#'],
      curSnap
    );
    expect({
      ...result.resultValue,
      '#': result.resultHam,
    }).toEqual({
      type: 'project',
      _id: 'cisvmclrpxin',
      '#': [
        1543234836586,
        {
          type: 1543234836586,
          _id: 1543234836586,
          name: 1543243812705,
          plans: [1543246278801, { cit4g6b5xqop: 1543246278801 }],
        },
      ],
      name: '1234',
      plans: {
        cit4g6b5xqop: {
          type: 'plan',
          _id: 'cit4g6b5xqop',
          name: 'plan 1543246278800',
        },
      },
    });
  });

  it('something with nested records', () => {
    const {
      machineState,
      incomingHam,
      incomingValue,
      currentHam,
      currentValue,
    } = {
      machineState: 1543263315259,
      incomingHam: [
        1543234836586,
        {
          type: 1543234836586,
          _id: 1543234836586,
          name: 1543243812705,
          plans: [
            1543246278801,
            { cit4g6b5xqop: [1543246278801, { '#': 1543263315073 }] },
          ],
          _rev: 1543262325753,
        },
      ] as HamValue,
      incomingValue: {
        type: 'project',
        _id: 'cisvmclrpxin',
        '#': [
          1543234836586,
          {
            type: 1543234836586,
            _id: 1543234836586,
            name: 1543243812705,
            plans: [
              1543246278801,
              { cit4g6b5xqop: [1543246278801, { '#': 1543263315073 }] },
            ],
            _rev: 1543262325753,
          },
        ],
        name: '1234',
        plans: {
          cit4g6b5xqop: {
            type: 'plan',
            _id: 'cit4g6b5xqop',
            name: 'plan 1543246278800',
          },
        },
      },
      currentHam: [
        1543234836586,
        {
          type: 1543234836586,
          _id: 1543234836586,
          name: 1543243812705,
          plans: [1543246278801, { cit4g6b5xqop: 1543246278801 }],
          _rev: 1543262325753,
        },
      ] as HamValue,
      currentValue: {
        _id: 'cisvmclrpxin',
        '#': [
          1543234836586,
          {
            type: 1543234836586,
            _id: 1543234836586,
            name: 1543243812705,
            plans: [1543246278801, { cit4g6b5xqop: 1543246278801 }],
            _rev: 1543262325753,
          },
        ],
        type: 'project',
        name: '1234',
        plans: {
          cit4g6b5xqop: {
            type: 'plan',
            _id: 'cit4g6b5xqop',
            name: 'plan 1543246278800',
          },
        },
        _rev: '3236-81275a1419d7075f47b79215e7e41791',
      },
    };
    expect(() =>
      merge(machineState, incomingHam, incomingValue, currentHam, currentValue)
    ).not.toThrow();
  });
});
