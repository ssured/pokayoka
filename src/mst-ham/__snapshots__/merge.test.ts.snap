// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`real world usage something with nested records 1`] = `
Object {
  "currentChanged": false,
  "resultHam": Array [
    1543234836586,
    Object {
      "_id": 1543234836586,
      "_rev": 1543262325753,
      "name": 1543243812705,
      "plans": Array [
        1543246278801,
        Object {
          "cit4g6b5xqop": Array [
            1543246278801,
            Object {
              "_id": 1543246278801,
              "name": 1543246278801,
              "type": 1543246278801,
            },
          ],
        },
      ],
      "type": 1543234836586,
    },
  ],
  "resultValue": Object {
    "#": Array [
      1543234836586,
      Object {
        "_id": 1543234836586,
        "_rev": 1543262325753,
        "name": 1543243812705,
        "plans": Array [
          1543246278801,
          Object {
            "cit4g6b5xqop": 1543246278801,
          },
        ],
        "type": 1543234836586,
      },
    ],
    "_id": "cisvmclrpxin",
    "_rev": "3236-81275a1419d7075f47b79215e7e41791",
    "name": "1234",
    "plans": Object {
      "cit4g6b5xqop": Object {
        "_id": "cit4g6b5xqop",
        "name": "plan 1543246278800",
        "type": "plan",
      },
    },
    "type": "project",
  },
}
`;
